Cluster Autoscaler: Segmentation violation caused by HintingSimulator predicateChecker nil pointer

I apologize in advance if this does not follow any issue template or any guidelines, but I think none of them fit.

I've created a [fork](https://github.com/carlocorradini/autoscaler) to support my own cloud provider called [*reCluster*](https://github.com/carlocorradini/reCluster). *reCluster* is based on [*K3s*](https://k3s.io). The implementation of [*recluster* cloudprovider](https://github.com/carlocorradini/autoscaler/tree/master/cluster-autoscaler/cloudprovider/recluster) for Cluster Autoscaler is "simple" and inspired by all other cloud providers.

When scaling down, Cluster Autoscaler performs flawlessly. Scale to 0 is supported, and the `controller` nodes are in a special node group (node pool) where scaling is disabled, like other cloud providers implement.

**ISSUE:**
When trying to **scale up** a segmentation violation occurs, crashing the Cluster Autoscaler. Strangely, **the violation is not done inside any of my (reCluster) code** but in `hinting_simulator.go` *findNode* function when trying to call `FitsAnyNodeMatching` on `predicateChecker` (https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/simulator/scheduling/hinting_simulator.go#L114).
`predicateChecker` is `nil` and I really don't know why...
Crash Log:
```
I1221 00:11:19.153088  148537 filter_out_schedulable.go:63] Filtering out schedulables
I1221 00:11:19.153126  148537 hinting_simulator.go:63] Looking for place for default/nginx-deployment-854db9b98b-jxb5d
I1221 00:11:19.153469  148537 round_trippers.go:463] GET https://192.168.0.157:6443/api/v1/namespaces/kube-system/configmaps/cluster-autoscaler-status
I1221 00:11:19.153502  148537 round_trippers.go:469] Request Headers:
I1221 00:11:19.153516  148537 round_trippers.go:473]     Accept: application/json, */*
I1221 00:11:19.153558  148537 round_trippers.go:473]     User-Agent: cluster-autoscaler-amd64/v0.0.0 (linux/amd64) kubernetes/$FormatI1221 00:11:19.159240  148537 round_trippers.go:574] Response Status: 200 OK in 5 milliseconds
I1221 00:11:19.159521  148537 round_trippers.go:463] PUT https://192.168.0.157:6443/api/v1/namespaces/kube-system/configmaps/cluster-autoscaler-status
I1221 00:11:19.159587  148537 round_trippers.go:469] Request Headers:
I1221 00:11:19.159603  148537 round_trippers.go:473]     Accept: application/json, */*
I1221 00:11:19.159642  148537 round_trippers.go:473]     User-Agent: cluster-autoscaler-amd64/v0.0.0 (linux/amd64) kubernetes/$FormatI1221 00:11:19.159660  148537 round_trippers.go:473]     Content-Type: application/json
I1221 00:11:19.163162  148537 round_trippers.go:574] Response Status: 200 OK in 3 milliseconds
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x17f1298]

goroutine 44 [running]:
k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling.(*HintingSimulator).findNode(0xc00049d3e0, 0xc000a9f190, {0x2096240, 0xc000136118}, 0xc0006c4c00, 0x1ea39e0)
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling/hinting_simulator.go:114 +0x338
k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling.(*HintingSimulator).TrySchedulePods(0x1997700?, {0x2096240, 0xc000136118}, {0xc000136600, 0x1, 0x1d31600?}, 0xc0006fe410?, 0x0)
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling/hinting_simulator.go:70 +0x2cf
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*filterOutSchedulablePodListProcessor).filterOutSchedulableByPacking(0xc000136128, {0xc000136600, 0x1, 0x1}, {0x2096240, 0xc000136118})
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/filter_out_schedulable.go:101 +0x111
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*filterOutSchedulablePodListProcessor).Process(0x0?, 0xc0006ec000, {0xc000136600?, 0x1, 0x1})
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/filter_out_schedulable.go:66 +0xd5
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*defaultPodListProcessor).Process(0xc000514200, 0xc0004e6c70?, {0xc000136600?, 0x2?, 0x2?})
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/pod_list_processor.go:45 +0x65
k8s.io/autoscaler/cluster-autoscaler/core.(*StaticAutoscaler).RunOnce(0xc00085ac80, {0x4?, 0xc0004e8f60?, 0x2fb23c0?})
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/core/static_autoscaler.go:477 +0x1ca3
main.run(0xc0000ead00?, {0x20913f8, 0xc000458660})
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/main.go:442 +0x2cd
main.main.func2({0x8aae06?, 0xc000160180?})
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/main.go:529 +0x25
created by k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
        [OMITTED]k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:211 +0x11b
```

If this is my oversight, I apologize. However, seeing that the segmentation violation is not present in any of my code warns me of a possible bug/misconfiguration.

The same issue with cluster-autoscaler v1.26.0, rancher v1.24.8+rke2r1 and Amazon cloud provider.  Cluster-autoscaler crashes during scale up attempt with following error:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x3cfca79]
goroutine 87 [running]:
k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling.(*HintingSimulator).findNode(0xc0009d2f30, 0x5a2fdc0?, {0x5a2fdc0, 0xc0000121c8}, 0xc000d06000, 0x0?)
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling/hinting_simulator.go:114 +0x159
k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling.(*HintingSimulator).TrySchedulePods(0x40a5c00?, {0x5a2fdc0, 0xc0000121c8}, {0xc000a9cbb0, 0x1, 0x4f2fca0?}, 0xc0006d8680?, 0x0)
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling/hinting_simulator.go:70 +0x2cf
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*filterOutSchedulablePodListProcessor).filterOutSchedulableByPacking(0xc0000121d8, {0xc000a9cbb0, 0x1, 0x1}, {0x5a2fdc0, 0xc0000121c8})
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/filter_out_schedulable.go:101 +0x111
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*filterOutSchedulablePodListProcessor).Process(0x1?, 0xc0013f6000, {0xc000a9cbb0?, 0x1, 0x1})
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/filter_out_schedulable.go:66 +0xd5
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*defaultPodListProcessor).Process(0xc000bfd240, 0xc00014f7c0?, {0xc000a9cbb0?, 0xa?, 0xa?})
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/pod_list_processor.go:45 +0x65
k8s.io/autoscaler/cluster-autoscaler/core.(*StaticAutoscaler).RunOnce(0xc000f628c0, {0x4?, 0x0?, 0x87e1e20?})
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/static_autoscaler.go:477 +0x1ca3
main.run(0x0?, {0x5a277b8, 0xc000b21350})
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:442 +0x2cd
main.main.func2({0x0?, 0x0?})
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:529 +0x25
created by k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:211 +0x11b

I guess that I have a similar issue:

After upgrading 1.25 to 1.26 I ran into this error. I checked and saw that I have 2 nodes in the terminating state.

```
I1227 14:39:06.965269       1 node_instances_cache.go:168] Refresh cloud provider node instances cache finished, refresh took 6.38µs
W1227 14:39:16.994547       1 clusterstate.go:428] AcceptableRanges have not been populated yet. Skip checking
W1227 14:39:16.994568       1 clusterstate.go:428] AcceptableRanges have not been populated yet. Skip checking
I1227 14:39:16.994671       1 static_autoscaler.go:388] 2 unregistered nodes present
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x3cfca79]

goroutine 103 [running]:
k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling.(*HintingSimulator).findNode(0xc000dc5830, 0x5a2fdc0?, {0x5a2fdc0, 0xc000012c08}, 0xc000c94c00, 0x0?)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling/hinting_simulator.go:114 +0x159
k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling.(*HintingSimulator).TrySchedulePods(0x40a5c00?, {0x5a2fdc0, 0xc000012c08}, {0xc003a6a888, 0x1, 0xc0005e5488?}, 0x451932?, 0x0)
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/simulator/scheduling/hinting_simulator.go:70 +0x2cf
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*filterOutSchedulablePodListProcessor).filterOutSchedulableByPacking(0xc000012c18, {0xc003a6a888, 0x1, 0x1}, {0x5a2fdc0, 0xc000012c08})
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/filter_out_schedulable.go:101 +0x111
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*filterOutSchedulablePodListProcessor).Process(0x0?, 0xc0004f7000, {0xc003a6a888?, 0x1, 0x1})
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/filter_out_schedulable.go:66 +0xd5
k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor.(*defaultPodListProcessor).Process(0xc000e88770, 0xc0042f4ae0?, {0xc003a6a888?, 0x6?, 0x6?})
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/podlistprocessor/pod_list_processor.go:45 +0x65
k8s.io/autoscaler/cluster-autoscaler/core.(*StaticAutoscaler).RunOnce(0xc000e45220, {0x4?, 0x0?, 0x87e1e20?})
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/static_autoscaler.go:477 +0x1ca3
main.run(0xc000ae9100?, {0x5a277b8, 0xc000a532f0})
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:442 +0x2cd
main.main.func2({0x8b9dc6?, 0xc000188a80?})
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:529 +0x25
created by k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
	/gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:211 +0x11b
```

--- a/cluster-autoscaler/core/autoscaler.go
+++ b/cluster-autoscaler/core/autoscaler.go
@@ -90,14 +90,6 @@ func initializeDefaultOptions(opts *AutoscalerOptions) error {
 	if opts.AutoscalingKubeClients == nil {
 		opts.AutoscalingKubeClients = context.NewAutoscalingKubeClients(opts.AutoscalingOptions, opts.KubeClient, opts.EventsKubeClient)
 	}
-	if opts.PredicateChecker == nil {
-		predicateCheckerStopChannel := make(chan struct{})
-		predicateChecker, err := predicatechecker.NewSchedulerBasedPredicateChecker(opts.KubeClient, predicateCheckerStopChannel)
-		if err != nil {
-			return err
-		}
-		opts.PredicateChecker = predicateChecker
-	}
 	if opts.ClusterSnapshot == nil {
 		opts.ClusterSnapshot = clustersnapshot.NewBasicClusterSnapshot()
 	}

--- a/cluster-autoscaler/main.go
+++ b/cluster-autoscaler/main.go
@@ -30,6 +30,7 @@ import (
 	"time"
 
 	"k8s.io/autoscaler/cluster-autoscaler/debuggingsnapshot"
+	"k8s.io/autoscaler/cluster-autoscaler/simulator/predicatechecker"
 
 	"github.com/spf13/pflag"
 
@@ -365,12 +366,18 @@ func buildAutoscaler(debuggingSnapshotter debuggingsnapshot.DebuggingSnapshotter
 	kubeClient := createKubeClient(getKubeConfig())
 	eventsKubeClient := createKubeClient(getKubeConfig())
 
+	predicateChecker, err := predicatechecker.NewSchedulerBasedPredicateChecker(kubeClient, make(chan struct{}))
+	if err != nil {
+		return nil, err
+	}
+
 	opts := core.AutoscalerOptions{
 		AutoscalingOptions:   autoscalingOptions,
 		ClusterSnapshot:      clustersnapshot.NewDeltaClusterSnapshot(),
 		KubeClient:           kubeClient,
 		EventsKubeClient:     eventsKubeClient,
 		DebuggingSnapshotter: debuggingSnapshotter,
+		PredicateChecker:     predicateChecker,
 	}
 
 	opts.Processors = ca_processors.DefaultProcessors()

